components/terms_of_service/terms_of_service.jsx

```diff
@@ -0,0 +1,259 @@
+
+import {browserHistory} from 'utils/browser_history';
+import messageHtmlToComponent from 'utils/message_html_to_component';
+import {formatText} from 'utils/text_formatting.jsx';
+import * as Utils from 'utils/utils.jsx';
+import {Constants} from 'utils/constants.jsx';
+
+export default class TermsOfService extends React.PureComponent {
+    static propTypes = {
+        customServiceTermsId: PropTypes.string.isRequired,
+        privacyPolicyLink: PropTypes.string,
+        siteName: PropTypes.string,
+        termsEnabled: PropTypes.bool,
+        termsOfServiceLink: PropTypes.string,
+    };
+
+    static defaultProps = {
+        privacyPolicyLink: 'https://about.mattermost.com/default-privacy-policy/',
+        siteName: 'Mattermost',
+        termsEnabled: true,
+        termsOfServiceLink: 'https://about.mattermost.com/default-terms/',
+    };
+
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            customServiceTermsText: '',
+            loading: true,
+            loadingAgree: false,
+            loadingDisagree: false,
+            serverError: null,
+        };
+
+        this.formattedText = memoizeResult((text) => formatText(text));
+    }
+
+    componentDidMount() {
+        if (this.props.termsEnabled) {
+            this.getServiceTerms();
+        } else {
+            GlobalActions.redirectUserToDefaultTeam();
+        }
+    }
+
+    componentDidUpdate(prevProps) {
+        if (prevProps.customServiceTermsId !== this.props.customServiceTermsId) {
+            this.getServiceTerms();
+        }
+    }
+
+    getServiceTerms = () => {
+        this.setState({
+            customServiceTermsText: '',
+            loading: false,
+        });
+        getServiceTerms(
+            (data) => {
+                this.setState({
+                    customServiceTermsText: data.text,
+                    loading: false,
+                });
+            },
+            () => {
+                GlobalActions.emitUserLoggedOutEvent(`/login?extra=${Constants.GET_TERMS_ERROR}`);
+                this.setState({
+                    loading: false,
+                });
+            }
+        );
+    };
+
+    handleLogoutClick = (e) => {
+        e.preventDefault();
+        GlobalActions.emitUserLoggedOutEvent('/login');
+    };
+
+    handleAcceptTerms = () => {
+        this.setState({
+            loadingAgree: true,
+            serverError: null,
+        });
+        this.registerUserAction(
+            'true',
+            () => {
+                const query = new URLSearchParams(this.props.location.search);
+                const redirectTo = query.get('redirect_to');
+                const redirectAction = query.get('redirect_action');
+                if (redirectTo && redirectTo.match(/^\/([^/]|$)/)) {
+                    browserHistory.push(redirectTo);
+                }
+                if (redirectAction === Constants.DEFAULT_TEAM_REDIRECT) {
```

Comment:
Would it make sense to just take this action if `redirectTo` was an empty string, avoiding the need to pass `redirectAction` at all -- just like we do below?

Reply:
This would be needed if we have another action that we want to perform apart from `redirectUserToDefaultTeam`.